File: python/섬의 개수(4964).py
from collections import deque

while True:
    w, h = map(int, input().split())

    if w == 0 and h == 0:
        break
    
    graph = [list(map(int, input().split())) for _ in range(h)]

    dx = [-1,1,0,0,-1,-1,1,1]
    dy = [0,0,-1,1,1,-1,1,-1]

    def bfs(x,y):
        queue = deque()
        queue.append([x,y])
        graph[x][y] = 2

        while queue:
            x, y = queue.popleft()

            for i in range(8):
                nx = x + dx[i]
                ny = y + dy[i]

                if nx < 0 or nx >= h or ny < 0 or ny >= w:
                    continue

                if graph[nx][ny] == 0:
                    continue

                if graph[nx][ny] == 1:
                    graph[nx][ny] = 2
                    queue.append([nx,ny])

    cnt = 0
    for i in range(h):
        for j in range(w):
            if graph[i][j] == 1:
                bfs(i,j)
                cnt += 1

    print(cnt)

File: js/섬의 개수(4964).js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

let index = 0;

while (true) {
    const [w, h] = input[index].split(' ').map(Number);
    index++;

    if (w === 0 && h === 0) break;
    const graph = [];
    for (let i=0; i<h; i++) {
        graph.push(input[index].split(' ').map(Number));
        index++;
    }

    const dx = [-1,1,0,0,-1,-1,1,1];
    const dy = [0,0,-1,1,1,-1,1,-1];

    function bfs(x,y) {
        const queue = [];
        queue.push([x,y]);
        graph[x][y] = 2;

        while (queue.length) {
            const [x,y] = queue.shift();

            for (let i=0; i<8; i++) {
                const nx = x + dx[i];
                const ny = y + dy[i];

                if (nx < 0 || nx >= h || ny < 0 || ny >= w) continue;
                if (graph[nx][ny] === 0) continue;
                if (graph[nx][ny] === 1) {
                    graph[nx][ny] = 2;
                    queue.push([nx,ny]);
                }
            }
        }
    }

    let cnt = 0;
    for (let i=0; i<h; i++) {
        for (let j=0; j<w; j++) {
            if (graph[i][j] === 1) {
                bfs(i,j);
                cnt++;
            }
        }
    }

    console.log(cnt);
}